google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,517 @@
+        if timing_header:
+            timing_data = SERVER_TIMING_REGEX.match(timing_header)
+            if timing_data and self.active_attempt:
+                self.active_attempt.gfe_latency_ms = float(timing_data.group(1))
+
+    @staticmethod
+    @lru_cache(maxsize=32)
+    def _parse_response_metadata_blob(blob: bytes) -> Tuple[str, str] | None:
+        """
+        Parse the response metadata blob and return a tuple of cluster and zone.
+
+        Function is cached to avoid parsing the same blob multiple times.
+
+        Args:
+          - blob: the metadata blob as extracted from the grpc call
+        Returns:
+          - a tuple of cluster_id and zone, or None if parsing failed
+        """
+        try:
+            proto = ResponseParams.pb().FromString(blob)
+            return proto.cluster_id, proto.zone_id
+        except (DecodeError, TypeError):
+            # failed to parse metadata
+            return None
+
+    def attempt_first_response(self) -> None:
+        """
+        Called to mark the timestamp of the first completed response for the
+        active attempt.
+
+        If the operation was completed, there is no active attempt, or the
+        active attempt already has a first response time, will raise an
+        exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if self.state != OperationState.ACTIVE_ATTEMPT or self.active_attempt is None:
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("attempt_first_response", self.state)
+            )
+        if self.active_attempt.first_response_latency_ms is not None:
+            return self._handle_error("Attempt already received first response")
+        # convert duration to milliseconds
+        self.active_attempt.first_response_latency_ms = (
+            time.monotonic() - self.active_attempt.start_time.monotonic
+        ) * 1000
+
+    def end_attempt_with_status(self, status: StatusCode | Exception) -> None:
+        """
+        Called to mark the end of a failed attempt for the operation.
+
+        If the operation was completed or there is no active attempt,
+        will raise an exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+
+        Args:
+          - status: The status of the attempt.
+        """
+        if self.state != OperationState.ACTIVE_ATTEMPT or self.active_attempt is None:
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("end_attempt_with_status", self.state)
+            )
+        if isinstance(status, Exception):
+            status = self._exc_to_status(status)
+        duration_seconds = time.monotonic() - self.active_attempt.start_time.monotonic
+        new_attempt = CompletedAttemptMetric(
+            start_time=self.active_attempt.start_time.utc,
+            first_response_latency_ms=self.active_attempt.first_response_latency_ms,
+            duration_ms=duration_seconds * 1000,
```

Comment:
is it possible to measure it in nano seconds? seconds precision seems too low :(

Reply:
duration_seconds is actually higher precision than seconds already, because it's a float value.

The docs [say this about the precision](https://docs.python.org/3/library/time.html#time.monotonic): "Use [monotonic_ns()](https://docs.python.org/3/library/time.html#time.monotonic_ns) to avoid the precision loss caused by the [float](https://docs.python.org/3/library/functions.html#float) type."

So I think we should already be at sub-millisecond percision, but if that's not good enough we can change everything to `monotonic_ns` to get full int nanoseconds everywhere

Reply:
Sure, just converted everything to int nanos